### ISSUE TITLE
BGP neighbor : Create/ Delete

### ISSUE THREAD

**Issue by jojoflyhigh** [fortinet-ansible-dev/ansible-galaxy-fortios-collection#98]
Hello,
I try to create a new neighbor BGP but I have the following message using this script below.
Second question to delete a neighbor using such script ?

     - name: Configure BGP Peer
      fortios_router_bgp:
        host: "X.X.X.X"
        username: "mylogin"
        password: "mypassword"
        ssl_verify: "False"
        vdom: "MY_VDOM"
        router_bgp:
          neighbor:
           - ip: '2.0.0.1'
             remote_as: "65111"

{
  "msg": "Error in repo",
  "meta": {
    "http_method": "POST",
    "status": "error",
    "http_status": 405,
    "vdom": "MY_VDOM",
    "path": "router",
    "name": "bgp",
    "serial": "FGmyserial",
    "version": "v6.2.6",
    "build": 6988
  },
  "invocation": {
    "module_args": {
      "host": "X.X.X.X",
      "username": "mylogin",
      "password": "mypassword",
      "ssl_verify": false,
      "vdom": "MY_VDOM",
      "router_bgp": {
        "neighbor": [
          {
            "ip": "2.0.0.1",
            "remote_as": "65111",
          }
        ],
        "admin_distance": null,
        "aggregate_address": null,
        "aggregate_address6": null,
        "always_compare_med": null,
        "as": null,
        "bestpath_as_path_ignore": null,
        "bestpath_cmp_confed_aspath": null,
        "bestpath_cmp_routerid": null,
        "bestpath_med_confed": null,
        "bestpath_med_missing_as_worst": null,
        "client_to_client_reflection": null,
        "cluster_id": null,
        "confederation_identifier": null,
        "confederation_peers": null,
        "dampening": null,
        "dampening_max_suppress_time": null,
        "dampening_reachability_half_life": null,
        "dampening_reuse": null,
        "dampening_route_map": null,
        "dampening_suppress": null,
        "dampening_unreachability_half_life": null,
        "default_local_preference": null,
        "deterministic_med": null,
        "distance_external": null,
        "distance_internal": null,
        "distance_local": null,
        "ebgp_multipath": null,
        "enforce_first_as": null,
        "fast_external_failover": null,
        "graceful_end_on_timer": null,
        "graceful_restart": null,
        "graceful_restart_time": null,
        "graceful_stalepath_time": null,
        "graceful_update_delay": null,
        "holdtime_timer": null,
        "ibgp_multipath": null,
        "ignore_optional_capability": null,
        "keepalive_timer": null,
        "log_neighbour_changes": null,
        "neighbor_group": null,
        "neighbor_range": null,
        "neighbor_range6": null,
        "network": null,
        "network_import_check": null,
        "network6": null,
        "redistribute": null,
        "redistribute6": null,
        "router_id": null,
        "scan_time": null,
        "synchronization": null
      },
      "https": true
    }
  },
  "_ansible_no_log": false,
  "changed": false
}

**Comment by JieX19** [comment]
Hi @jojoflyhigh,

You're using the legacy mode which is deprecated already. You can install our latest ansible galaxy collection by the following command:

**ansible-galaxy collection install fortinet.fortios:1.1.9**

Doc page https://ansible-galaxy-fortios-docs.readthedocs.io/en/latest/gen/fortios_router_bgp.html

The playbook should be like this: 
```
  collections:
    - fortinet.fortios
  connection: httpapi
  vars:
   vdom: "root"
   ansible_httpapi_use_ssl: yes
   ansible_httpapi_validate_certs: no
   ansible_httpapi_port: 443
  tasks:
  - name: Configure BGP.
    fortios_router_bgp:
      vdom:  "{{ vdom }}"
      router_bgp:
        neighbor:
              
```

It's not allowed to delete any global objects, router_bgp is one of them. You can modify the attributes if needed.

**Comment by jojoflyhigh** [comment]
Hello JieX19 ,
I was using fortinet.fortios:1.1.6 as my fortinet is in 6.2.6 release. 
I've tried with 1.1.9 module with the same unsuccess result.

**Comment by JieX19** [comment]
@jojoflyhigh,

Can you copy your entire playbook for me? I cannot reproduce the error on my side.

**Comment by jojoflyhigh** [comment]
Hello @JieX19, I'm using ansible AWX.
I've installed fortinet.fortios:1.1.9 in the awx_task docker container in order to support fortinet devices.
Here is my playbook :

---
- hosts: Fortinet
  gather_facts: no
  connection: httpapi
  environment:
    NO_PROXY: '10.194.138.130/32' 
  vars:
    vdom: "root"
    ansible_httpapi_use_ssl: yes
    ansible_httpapi_validate_certs: no
    ansible_httpapi_port: 443
  tasks:
    - name: Configure BGP Peer
      fortios_router_bgp:
        host: "10.194.138.130"
        username: "admin"
        password: ""
        ssl_verify: "False"
        vdom: "MY_VDOM"
        router_bgp:
          neighbor:
           - ip: '2.0.0.1'
             remote_as: "65111"
             bfd: "enable"
             send_community6: "disable"

**Comment by JieX19** [comment]
@jojoflyhigh,

Got it. The error is basically caused by the playbook. We're not using host, username, password, ssl_verify these variables in the ansible galaxy collection anymore. Instead, we're using ansible_httpapi_use_ssl, ansible_httpapi_validate_certs and ansible_httpapi_port.

Your playbook should works fine by deleting the variables mentioned above.

**Comment by jojoflyhigh** [comment]
@JieX19,
I've remove host, username, password, ssl_verify but I can't connect anymore to my node.

 /ansible/module_utils/connection.py\", line 185, in __rpc__\nansible.module_utils.connection.ConnectionError: Could not connect to https://10.194.138.130:443/logincheck: Tunnel connection failed: 403 Forbidden\n",
  "msg": "MODULE FAILURE\n"

What I did not understand is that I'm using that same syntaxe to configure BGP and redistibuste static routes into BGP without issue.

My host is configured as follow :
---
ansible_host: 10.194.138.130
ansible_connection: httpapi
ansible_network_os: fortinet.fortios.fortios
ansible_user: ansible
ansible_password: password

**Comment by JieX19** [comment]
@jojoflyhigh,

It seems the httpapi is not connected correctly. Also you didn't assign **collections** in your playbook, which will cause another error (cannot find the module you're calling) in the next step. The var **collections** is assigned a value **fortinet.fortios**. 

1. Configure the host file (use access_token if you have generated on from GUI)
[fortigate]
fortigate01 ansible_host=10.194.138.130 ansible_user="admin" ansible_password="password" access_token="abcdefghijklmn"
[fortigate:vars]
ansible_network_os=fortinet.fortios.fortios

2. Playbook 
```
- hosts: Fortinet
  connection: httpapi
  collections:
    - fortinet.fortios
  vars:
   vdom: "root"
   ansible_httpapi_use_ssl: yes
   ansible_httpapi_validate_certs: no
   ansible_httpapi_port: 443
  tasks:
  - name: Configure BGP.
    fortios_router_bgp:
      vdom:  "{{ vdom }}"
      router_bgp:
        neighbor:
```

**Comment by jojoflyhigh** [comment]
Hello @JieX19 

Unfortunately I did not succeed in the creation of BGP neighbor in release 2.0.0 and 2.0.1
What looks stange is that TASK [Gathering Facts] connection is OK but my own tack failed to connect.
Thanks for your help

Here is my playbook : 
---
- hosts: FTG1800F-1
  collections:
   - fortinet.fortios
  connection: httpapi
  vars:
   vdom: "root"
   ansible_httpapi_use_ssl: yes
   ansible_httpapi_validate_certs: no
   ansible_httpapi_port: 443
  environment:
    NO_PROXY: '10.194.138.130/32'
  tasks:

    - name: Configure BGP Peer
      fortios_router_bgp:
        vdom: "TREX_V"
        access_token: "3N5bj39p0ygywx54pzHn7xz7m5spQG"
        router_bgp:
          neighbor:
           - ip: '2.0.0.1'
             remote_as: "65111"
             bfd: "enable"
             send_community6: "disable"


And here is my logs : 
ansible-playbook 2.9.5
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/var/lib/awx/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /var/lib/awx/venv/ansible/lib/python3.6/site-packages/ansible
  executable location = /var/lib/awx/venv/ansible/bin/ansible-playbook
  python version = 3.6.8 (default, Aug 24 2020, 17:57:11) [GCC 8.3.1 20191121 (Red Hat 8.3.1-5)]
Using /etc/ansible/ansible.cfg as config file
setting up inventory plugins
host_list declined parsing /tmp/awx_2056_fm4iyljv/tmpij2grmml as it did not pass its verify_file() method
Parsed /tmp/awx_2056_fm4iyljv/tmpij2grmml inventory source with script plugin
[
Loading callback plugin awx_display of type stdout, v2.0 from /var/lib/awx/venv/awx/lib/python3.6/site-packages/ansible_runner/callbacks/awx_display.py
[

PLAYBOOK: create_Fortinet_BGP.yml **********************************************
Positional arguments: create_Fortinet_BGP.yml
verbosity: 4
remote_user: root
connection: smart
timeout: 10
become_method: sudo
tags: ('all',)
inventory: ('/tmp/awx_2056_fm4iyljv/tmpij2grmml',)
extra_vars: ('@/tmp/awx_2056_fm4iyljv/env/extravars',)
forks: 5
[
1 plays in create_Fortinet_BGP.yml
[

PLAY [FTG1800F-1] **************************************************************

TASK [Gathering Facts] *********************************************************
task path: /tmp/awx_2056_fm4iyljv/project/create_Fortinet_BGP.yml:3
[
<10.194.138.130> attempting to start connection
[
<10.194.138.130> using connection plugin httpapi
[
<10.194.138.130> local domain socket does not exist, starting it
[
<10.194.138.130> control socket path is /var/lib/awx/.ansible/pc/db9bd82c0b
[
<10.194.138.130> local domain socket listeners started successfully
[
<10.194.138.130> loaded API plugin ansible_collections.fortinet.fortios.plugins.httpapi.fortios from path /var/lib/awx/.ansible/collections/ansible_collections/fortinet/fortios/plugins/httpapi/fortios.py for network_os fortinet.fortios.fortios
[
<10.194.138.130> 
[
<10.194.138.130> local domain socket path is /var/lib/awx/.ansible/pc/db9bd82c0b
[
<10.194.138.130> ESTABLISH LOCAL CONNECTION FOR USER: root
[
<10.194.138.130> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867741.8405528-265333158403328 `" && echo ansible-tmp-1617867741.8405528-265333158403328="` echo /var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867741.8405528-265333158403328 `" ) && sleep 0'
[
<FTG1800F-1> Attempting python interpreter discovery
[
<10.194.138.130> EXEC /bin/sh -c 'echo PLATFORM; uname; echo FOUND; command -v '"'"'/usr/bin/python'"'"'; command -v '"'"'python3.7'"'"'; command -v '"'"'python3.6'"'"'; command -v '"'"'python3.5'"'"'; command -v '"'"'python2.7'"'"'; command -v '"'"'python2.6'"'"'; command -v '"'"'/usr/libexec/platform-python'"'"'; command -v '"'"'/usr/bin/python3'"'"'; command -v '"'"'python'"'"'; echo ENDFOUND && sleep 0'
[
<10.194.138.130> EXEC /bin/sh -c '/var/lib/awx/venv/ansible/bin/python3.6 && sleep 0'
[
Using module file /var/lib/awx/venv/ansible/lib/python3.6/site-packages/ansible/modules/system/setup.py
[
<10.194.138.130> PUT /var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/tmp697spbr5 TO /var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867741.8405528-265333158403328/AnsiballZ_setup.py
[
<10.194.138.130> EXEC /bin/sh -c 'chmod u+x /var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867741.8405528-265333158403328/ /var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867741.8405528-265333158403328/AnsiballZ_setup.py && sleep 0'
[
<10.194.138.130> EXEC /bin/sh -c 'NO_PROXY=10.194.138.130/32 /usr/libexec/platform-python /var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867741.8405528-265333158403328/AnsiballZ_setup.py && sleep 0'
[
<10.194.138.130> EXEC /bin/sh -c 'rm -f -r /var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867741.8405528-265333158403328/ > /dev/null 2>&1 && sleep 0'
[
ok: [FTG1800F-1]
[
META: ran handlers
[

TASK [Configure BGP Peer] ******************************************************
task path: /tmp/awx_2056_fm4iyljv/project/create_Fortinet_BGP.yml:16
[
<10.194.138.130> attempting to start connection
[
<10.194.138.130> using connection plugin httpapi
[
<10.194.138.130> found existing local domain socket, using it!
[
<10.194.138.130> updating play_context for connection
[
<10.194.138.130> 
[
<10.194.138.130> local domain socket path is /var/lib/awx/.ansible/pc/db9bd82c0b
[
<10.194.138.130> ESTABLISH LOCAL CONNECTION FOR USER: root
[
<10.194.138.130> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867743.9298306-136581746124583 `" && echo ansible-tmp-1617867743.9298306-136581746124583="` echo /var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867743.9298306-136581746124583 `" ) && sleep 0'
[
Using module file /var/lib/awx/.ansible/collections/ansible_collections/fortinet/fortios/plugins/modules/fortios_router_bgp.py
[
<10.194.138.130> PUT /var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/tmpmgirvebu TO /var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867743.9298306-136581746124583/AnsiballZ_fortios_router_bgp.py
[
<10.194.138.130> EXEC /bin/sh -c 'chmod u+x /var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867743.9298306-136581746124583/ /var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867743.9298306-136581746124583/AnsiballZ_fortios_router_bgp.py && sleep 0'
[
<10.194.138.130> EXEC /bin/sh -c 'NO_PROXY=10.194.138.130/32 /usr/libexec/platform-python /var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867743.9298306-136581746124583/AnsiballZ_fortios_router_bgp.py && sleep 0'
[
<10.194.138.130> EXEC /bin/sh -c 'rm -f -r /var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867743.9298306-136581746124583/ > /dev/null 2>&1 && sleep 0'
[
The full traceback is:
Traceback (most recent call last):
  File "/var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867743.9298306-136581746124583/AnsiballZ_fortios_router_bgp.py", line 102, in <module>
    _ansiballz_main()
  File "/var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867743.9298306-136581746124583/AnsiballZ_fortios_router_bgp.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867743.9298306-136581746124583/AnsiballZ_fortios_router_bgp.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible_collections.fortinet.fortios.plugins.modules.fortios_router_bgp', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib64/python3.6/runpy.py", line 205, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_fortios_router_bgp_payload_54q1gxo5/ansible_fortios_router_bgp_payload.zip/ansible_collections/fortinet/fortios/plugins/modules/fortios_router_bgp.py", line 10404, in <module>
  File "/tmp/ansible_fortios_router_bgp_payload_54q1gxo5/ansible_fortios_router_bgp_payload.zip/ansible_collections/fortinet/fortios/plugins/modules/fortios_router_bgp.py", line 10381, in main
  File "/tmp/ansible_fortios_router_bgp_payload_54q1gxo5/ansible_fortios_router_bgp_payload.zip/ansible_collections/fortinet/fortios/plugins/module_utils/fortios/fortios.py", line 248, in check_schema_versioning
  File "/tmp/ansible_fortios_router_bgp_payload_54q1gxo5/ansible_fortios_router_bgp_payload.zip/ansible/module_utils/connection.py", line 185, in __rpc__
ansible.module_utils.connection.ConnectionError: Could not connect to https://10.194.138.130:443/logincheck?access_token=3N5bj39p0ygywx54pzHn7xz7m5spQG: Tunnel connection failed: 403 Forbidden
fatal: [FTG1800F-1]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\\n  File \\"/var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867743.9298306-136581746124583/AnsiballZ_fortios_router_bgp.py\\", line 102, in <module>\\n    _ansiballz_main()\\n  File \\"/var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867743.9298306-136581746124583/AnsiballZ_fortios_router_bgp.py\\", line 94, in _ansiballz_main\\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\\n  File \\"/var/lib/awx/.ansible/tmp/ansible-local-133258c328ppoi/ansible-tmp-1617867743.9298306-136581746124583/AnsiballZ_fortios_router_bgp.py\\", line 40, in invoke_module\\n    runpy.run_module(mod_name='ansible_collections.fortinet.fortios.plugins.modules.fortios_router_bgp', init_globals=None, run_name='__main__', alter_sys=True)\\n  File \\"/usr/lib64/python3.6/runpy.py\\", line 205, in run_module\\n    return _run_module_code(code, init_globals, run_name, mod_spec)\\n  File \\"/usr/lib64/python3.6/runpy.py\\", line 96, in _run_module_code\\n    mod_name, mod_spec, pkg_name, script_name)\\n  File \\"/usr/lib64/python3.6/runpy.py\\", line 85, in _run_code\\n    exec(code, run_globals)\\n  File \\"/tmp/ansible_fortios_router_bgp_payload_54q1gxo5/ansible_fortios_router_bgp_payload.zip/ansible_collections/fortinet/fortios/plugins/modules/fortios_router_bgp.py\\", line 10404, in <module>\\n  File \\"/tmp/ansible_fortios_router_bgp_payload_54q1gxo5/ansible_fortios_router_bgp_payload.zip/ansible_collections/fortinet/fortios/plugins/modules/fortios_router_bgp.py\\", line 10381, in main\\n  File \\"/tmp/ansible_fortios_router_bgp_payload_54q1gxo5/ansible_fortios_router_bgp_payload.zip/ansible_collections/fortinet/fortios/plugins/module_utils/fortios/fortios.py\\", line 248, in check_schema_versioning\\n  File \\"/tmp/ansible_fortios_router_bgp_payload_54q1gxo5/ansible_fortios_router_bgp_payload.zip/ansible/module_utils/connection.py\\", line 185, in __rpc__\\nansible.module_utils.connection.ConnectionError: Could not connect to https://10.194.138.130:443/logincheck?access_token=3N5bj39p0ygywx54pzHn7xz7m5spQG: Tunnel connection failed: 403 Forbidden\\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\\nSee stdout/stderr for the exact error",
    "rc": 1
}
[

PLAY RECAP *********************************************************************
FTG1800F-1                 : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

**Comment by chillancezen** [comment]
hi @jojoflyhigh,

the error says it can not connect:
```
ansible.module_utils.connection.ConnectionError: Could not connect to https://10.194.138.130:443/logincheck?access_token=3N5bj39p0ygywx54pzHn7xz7m5spQG: Tunnel connection failed: 403 Forbidden
```

why do you put these lines there?
```
  environment:
    NO_PROXY: '10.194.138.130/32'
```

could you have a try without them?

**Comment by chillancezen** [comment]
below are mine, and they ran well.
```
$cat hosts
[fortigates]
fortigate03 ansible_host=192.168.190.171 fortios_access_token=<mytoken>  ansible_user="admin" ansible_password="<mypassword>"

[fortigates:vars]
ansible_network_os=fortinet.fortios.fortios
```

```
$cat router_bgp.yml
- hosts: fortigate03
  connection: httpapi
  collections:
  - fortinet.fortios
  vars:
   vdom: "root"
   ansible_httpapi_use_ssl: yes
   ansible_httpapi_validate_certs: no
   ansible_httpapi_port: 443
  tasks:
   -  name: Enable threat feed on AV profile
      fortinet.fortios.fortios_router_bgp:
        access_token: "{{ fortios_access_token }}"
        router_bgp:
            neighbor:
                - ip: '2.0.0.1'
                  remote_as: '65111'
```

**Comment by jojoflyhigh** [comment]
@JieX19, @chillancezen 

Hello, I finally succeed in configuring BGP neighbor by re-installing ansible after a cleanup. It was due to two things that maybe can help somebody else:
1) I think that I was using deprecated module even if the new one was installed.
2) Proxy issue. Requests where going through http_proxy like with deprecated module. But it didn't work properly with the new module. Configuring no_proxy for my host has resolved this issue.

Thanks for your help
